feat(checks): add locale test

smerle33 opened the pull request:
as per https://github.com/jenkins-infra/helpdesk/issues/3138 implementation of the first checks for linux.

smerleCB committed:
feat(checks): add few test cf https://github.com/jenkins-infra/helpdesk/issues/3138

dduportal commented on Jenkinsfile:
```diff
@@ -27,20 +28,25 @@ def generateParallelSteps(labels) {
+                    //         cat /proc/meminfo
+                    //     fi
+                    // '''
+                    def localeOutput = sh "locale -a"
```
It is recommended to use shell steps instead of pure pipeline for this kind of logic

smerleCB committed:
remove other agent

smerleCB committed:
add an agent

smerleCB committed:
add onpush trigger

smerleCB committed:
R & D

smerleCB committed:
maven

smerleCB committed:
remove windows test for now

smerleCB committed:
add maven check

smerleCB committed:
remove most of the agent for now

smerleCB committed:
add all linux agents for check

smerleCB committed:
simplify ok message for maven

smerleCB committed:
check datadog journalctl

smerleCB committed:
check datadog journalctl

smerleCB committed:
check datadog agent

smerleCB committed:
check datadog agent health

smerleCB committed:
check datadog agent version

smerleCB committed:
check datadog agent version AND service status

smerleCB committed:
replan all the agents to check

smerleCB committed:
first set of checks

smerleCB committed:
chore typo

dduportal commented on checks.sh:
```diff
@@ -1,89 +1,84 @@
-# fi
+set +u
+if [[ -z "${JAVA_HOME}" ]]; then
+    echo "ERROR JDK environement variable not defined"
```
```suggestion
    echo "ERROR: the environment variable 'JAVA_HOME' is undefined"
```

dduportal commented on checks.sh:
```diff
@@ -1,89 +1,84 @@
+    elif [ "$1" = "kubernetes" ]; then
+        jdk="jdk-8"
+    fi
+    if [[ ${JAVA_HOME} != *"$jdk"* ]]; then
```
It's an acceptance test: no need to "finely" test : the output of `mvn -v` should contain the string `$jdk`

```suggestion
if [[ "$(mvn -v 2>&1)" != *"$jdk"* ]]; then
```

smerle33 commented:
<img width="1487" alt="Capture d’écran 2022-09-23 à 09 03 03" src="https://user-images.githubusercontent.com/95630726/191909104-ad1a0a58-a261-4616-a25c-677ea15293ee.png">

smerleCB committed:
reformat bash

lemeurherve commented on checks.sh:
```diff
@@ -45,14 +45,14 @@ fi
 set +u
 if [[ -z "${JAVA_HOME}" ]]; then
-    echo "ERROR JDK environement variable not defined"
+    echo "ERROR: the environment variable 'JAVA_HOME' is undefined"
```
```suggestion
    echo "ERROR: the 'JAVA_HOME' environment variable is undefined"
```

lemeurherve commented on checks.sh:
```diff
@@ -18,10 +18,10 @@ if test -e /proc/meminfo; then
+    echo "${DefaultLocale} locale is available"
 else
-    echo "ERROR $DefaultLocale locale is not available $(locale -a)"
+    echo "ERROR ${DefaultLocale} locale is not available $(locale -a)"
```
```suggestion
    echo "ERROR: ${DefaultLocale} locale is not available $(locale -a)"
```

lemeurherve commented on checks.sh:
```diff
@@ -64,21 +64,21 @@ if [ $# -ge 1 ] && [ -n "$1" ]; then
-        echo "expecting $jdk for label $1"
-        echo "found ${JAVA_HOME}"
+    if [[ "$(mvn -v 2>&1)" != *"${jdk}"* ]]; then
+        echo "ERROR JDK not matching what is expected: "
+        echo "expecting ${jdk} for label $1"
```
```suggestion
        echo "ERROR: JDK not matching the expected ${jdk} for label '$1'"
```

lemeurherve commented on checks.sh:
```diff
@@ -64,21 +64,21 @@ if [ $# -ge 1 ] && [ -n "$1" ]; then
         exit 1
     else
-        echo "JDK ok : ${JAVA_HOME} for $1"
+        echo "JDK ok ${jdk} for $1"
```
```suggestion
        echo "${jdk} OK for label '$1'"
```

lemeurherve commented on checks.sh:
```diff
@@ -64,21 +64,21 @@ if [ $# -ge 1 ] && [ -n "$1" ]; then
     exit 1
 else
-    echo "MAVEN ok : $DefaultMavenVersion for $1"
+    echo "MAVEN ok: ${DefaultMavenVersion} for $1"
```
```suggestion
    echo "Maven version ${DefaultMavenVersion} OK for label '$1'"
```

lemeurherve commented on checks.sh:
```diff
@@ -64,21 +64,21 @@ if [ $# -ge 1 ] && [ -n "$1" ]; then
     echo "ERROR Maven version not matching what is expected : "
-    echo "expecting $DefaultMavenVersion for label $1"
-    echo "found $(mvn -v)"
+    echo "expecting ${DefaultMavenVersion} for label $1"
```
```suggestion
    echo "expecting ${DefaultMavenVersion} for label '$1'"
```

lemeurherve reviewed (commented):
Nits only, good work!

lemeurherve commented on checks.sh:
```diff
@@ -0,0 +1,84 @@
+fi
+
+if getent passwd jenkins >/dev/null; then
+    echo "jenkins user exists"
+else
+    echo "ERROR jenkins user does not exist"
```
```suggestion
    echo "'jenkins' user exists"
else
    echo "ERROR: 'jenkins' user does not exist"
```

lemeurherve commented on checks.sh:
```diff
@@ -0,0 +1,84 @@
+fi
+
+if [[ "$(whoami)" != "jenkins" ]]; then
+    echo "ERROR Not running as jenkins user"
+    echo "whoami: $(whoami)"
+    exit 1
+fi
+
+if sudo -n whoami; then
+    echo "ERROR running as root possible"
```
```suggestion
    echo "ERROR: Not running as 'jenkins' user"
    echo "whoami: $(whoami)"
    exit 1
fi

if sudo -n whoami; then
    echo "ERROR: running as root should not be possible"
```

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

smerle33 committed:
Update checks.sh

Co-authored-by: Hervé Le Meur <91831478+lemeurherve@users.noreply.github.com>

MarkEWaite commented on Jenkinsfile:
```diff
@@ -12,12 +12,13 @@ properties([
+sequentialStages['OS & Java'] = [ 'linux', 'jdk8', 'jdk11', 'windows']
 sequentialStages['Processor'] = [ 'arm64', 'amd64' ] // Remove ppc64le and s390x until virtual machine available 'ppc64le', 's390x'
-sequentialStages['Docker'] = [ 'arm64docker', 'docker', 'docker-windows' ] // Remove ppc64le and s390x until available again 'ppc64ledocker', 's390xdocker'
+sequentialStages['Docker'] = [ 'arm64docker', 'docker', 'windows'] // Remove ppc64le and s390x until available again 'ppc64ledocker', 's390xdocker'
```
As far as I understand it, we need a label that means "docker for Windows containers".  I think that we should be checking 

* "docker for aarch64 linux containers" with the "arm64docker" label
* "docker for amd64 linux containers" with the "docker" label
* "docker for amd64 windows containers"  with the "docker-windows" label
```suggestion
sequentialStages['Docker'] = [ 'arm64docker', 'docker', 'docker-windows'] // Remove ppc64le and s390x until available again 'ppc64ledocker', 's390xdocker'
```

I think this change is required, otherwise we will stop checking that we can allocate containers for Windows.

smerle33 replied:
this is clearly a mistake, I didn't plan to change the agent list ...

MarkEWaite committed:
Make checks.sh executable

MarkEWaite commented on Jenkinsfile:
```diff
@@ -27,18 +28,8 @@ def generateParallelSteps(labels) {
-                        fi
-                    '''
+                    checkout scm
+                    sh 'bash ./checks.sh '+label
```
Why not make the script executable and call it without the invocation of `bash`?  Is there a concern that we may run FreeBSD or OpenBSD or macOS agents where `/bin/bash` does not necessarily exist?

This suggested change is optional.  You're welcome to reject it if you would rather not make the file executable.
```suggestion
                    sh './checks.sh '+label
```

smerle33 replied:
if I remember correctly, the main concern is for the file rights (execution) and then the fact of potential nonexistence of bash as you mentioned.

dduportal replied:
What @smerle33 is correct: `bash` is a requirement for the script (as per its shebang at least) so it's now explicit in the pipeline (Particularly since the `sh` pipeline instruction is using `sh` by default) call and the execution bit is not always guaranteed to be applied on the shell script (depends on how it is retrieved and on the underlying filesystem).

MarkEWaite committed:
Add a comment on the java check

It checks the path of the Java runtime reported by `mvn -v`.  If
the path includes the expected string, the test passes.  If the path
does not include the expected string, the test fails.

I had been expecting a test that checks the java version number as
reported by `mvn -v`.

MarkEWaite committed:
Run all tests even if one test fails

Report failure at end of all tests rather than exit at first test.

MarkEWaite reviewed (changes requested):
Thanks very much!  A very nice improvement!

I've pushed two changes to the branch that you are welcome to revert or force push away if they don't meet the needs.  The changes that I pushed are:

* c0e1fe119bc55ca09760daf75a6d012961143697 - comment that the Java check is checking the path to the Java executable as reported by `mvn -v`, not the Java version as reported by `mvn -v`
* 44896fe6ec4febef3d817f8cc222602292ab74a1 - run all tests and report failures at the end rather than stopping all tests on the first failure

Apologies if the push is disruptive, but I couldn't see an easy way to express the second through suggested changes in GitHub.

smerle33 committed:
Apply suggestions from code review

Co-authored-by: Mark Waite <mark.earl.waite@gmail.com>

smerleCB committed:
add a new java version check from maven output

smerleCB committed:
chore missing $

smerleCB committed:
eval

smerleCB committed:
extract jdk number

smerleCB committed:
chore comment

smerle33 commented:
> Thanks very much! A very nice improvement!
> 
> I've pushed two changes to the branch that you are welcome to revert or force push away if they don't meet the needs. The changes that I pushed are:
> 
>     * [c0e1fe1](https://github.com/jenkins-infra/acceptance-tests/commit/c0e1fe119bc55ca09760daf75a6d012961143697) - comment that the Java check is checking the path to the Java executable as reported by `mvn -v`, not the Java version as reported by `mvn -v`
> 
>     * [44896fe](https://github.com/jenkins-infra/acceptance-tests/commit/44896fe6ec4febef3d817f8cc222602292ab74a1) - run all tests and report failures at the end rather than stopping all tests on the first failure
> 
> 
> Apologies if the push is disruptive, but I couldn't see an easy way to express the second through suggested changes in GitHub.

@MarkEWaite what do you think of ? : https://github.com/jenkins-infra/acceptance-tests/blob/3dd78eecd4688394743816ffb69faf083ba962a1/checks.sh#L82-L124

MarkEWaite commented:
> @MarkEWaite what do you think of ? :

That works great for me.  Thanks for implementing it.

MarkEWaite reviewed (approved):
Looks great to me.  Let's merge it and get the benefit from it.  Thanks!
